### PR TITLE
add: #22006 on USDT

### PR DESCRIPTION
**Pull request by 0xB10C**
Questions follow soon <sup><sup><sup><sup><sup><sup>:tm:</sup> </sup> </sup> </sup> </sup> </sup>

### CODE REVIEW

**Comment by 0xB10C** [comment]
image

forgot that ruby doesn't like my pseudonym.. pushing a fix

File: _posts/2021-05-26-#22006.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+[systemtaps]: https://sourceware.org/systemtap/wiki
+[#19866]: https://github.com/bitcoin/bitcoin/pull/19866
+[#20981]: https://github.com/bitcoin/bitcoin/issues/20981
+
+### Tracepoints and tracing scripts
+
+- We can hook into the tracepoints with tracing scripts via the Linux kernel.
+- If we don't hook into the tracepoints, then they are [NOP]s and have little to no performance impact.
+- The tracepoints can pass data back to the tracing script, which contains the tracing logic (just printing the data, collecting statistics, alerting, visualization, ...).
+- Tracepoints need to be somewhat generic to allow for reusability in different tracing scripts, but they also need a clear use case. There is no need in plastering the code with (unused) tracepoints.
```
Comment (jonatack) [inline _posts/2021-05-26-#22006.md:30]:
```suggestion
- Tracepoints need to be somewhat generic to allow for reusability in different tracing scripts, but they also need a clear use case. There is no need to plaster the code with (unused) tracepoints.
```

File: _posts/2021-05-26-#22006.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+[systemtaps]: https://sourceware.org/systemtap/wiki
+[#19866]: https://github.com/bitcoin/bitcoin/pull/19866
+[#20981]: https://github.com/bitcoin/bitcoin/issues/20981
+
+### Tracepoints and tracing scripts
+
+- We can hook into the tracepoints with tracing scripts via the Linux kernel.
+- If we don't hook into the tracepoints, then they are [NOP]s and have little to no performance impact.
+- The tracepoints can pass data back to the tracing script, which contains the tracing logic (just printing the data, collecting statistics, alerting, visualization, ...).
```
Comment (jonatack) [inline _posts/2021-05-26-#22006.md:29]:
```suggestion
- The tracepoints can pass data back to the tracing script, which contains the tracing logic, e.g. to collect statistics, print or visualize data, give alerts, etc.
```

File: _posts/2021-05-26-#22006.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+- Tracepoints need to be somewhat generic to allow for reusability in different tracing scripts, but they also need a clear use case. There is no need in plastering the code with (unused) tracepoints.
+- There are currently two main tools for writing USDT scripts. [bpftrace] for one-liners and short scripts and [BPF compiler collection (BCC)] e.g. with the Python wrapper for complex scripts. Other tools with USDT support are under development.
```
Comment (jonatack) [inline _posts/2021-05-26-#22006.md:31]: s/scripts./scripts:/

and maybe make a list
```diff
-- There are currently two main tools for writing USDT scripts. [bpftrace] for one-liners and short scripts and [BPF compiler collection (BCC)] e.g. with the Python wrapper for complex scripts. Other tools with USDT support are under development.
+- There are currently two main tools for writing USDT scripts:
+    - [bpftrace] for one-liners and short scripts
+    - [BPF compiler collection (BCC)], e.g. with the Python wrapper for complex scripts
+    - Other tools with USDT support are under development.
```

File: _posts/2021-05-26-#22006.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    Space    & loads │      │ tracing script  │                    │
+    ─────────────────┼──────┼─────────────────┼────────────────────┼───
+    Kernel           │      │                 │                    │
+    Space       ┌──┬─▼──────┴─────────────────┴────────────┐       │
+                │  │  eBPF program                         │◄──────┘
+                │  └───────────────────────────────────────┤
+                │ eBPF kernel Virtual Machine (sandboxed)  │
+                └──────────────────────────────────────────┘
+
+1. The tracing script complies the eBPF code and loads the eBFP program into a kernel VM
```
Comment (jonatack) [inline _posts/2021-05-26-#22006.md:64]: compiles

File: _posts/2021-05-26-#22006.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+2. The eBPF program hooks into one or more tracepoints
+3. When the tracepoint is called, the arguments are passed to the eBPF program
+4. The eBPF program processes the arguments and returns data to the tracing script
+```
+
+- More information on eBPF:
```
Comment (jonatack) [inline _posts/2021-05-26-#22006.md:70]:
```suggestion
More information on eBPF:
```

File: _posts/2021-05-26-#22006.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+- [collabora.com: An eBPF overview, part 2: Machine & bytecode](https://www.collabora.com/news-and-blog/blog/2019/04/15/an-ebpf-overview-part-2-machine-and-bytecode/)
+- [Brendan D. Gregg's blog posts, and book on on eBPF](http://www.brendangregg.com/)
+- [Brendan D. Gregg: Linux bcc/BPF Node.js USDT Tracing](http://www.brendangregg.com/blog/2016-10-12/linux-bcc-nodejs-usdt.html)
+
+
+
+### Examples
+
+- The PR includes examples and documentation on how to run them.
+- For building Bitcoin Core with USDT support, you need the `sys/std.h` headers (when present, USDT support is [automatically] compiled in). On Debian-like systems you can install them with `apt install systemtap-sdt-dev` (this is not documented yet).
```
Comment (jonatack) [inline _posts/2021-05-26-#22006.md:81]: I'm not sure what "you can install them" refers to here, installing headers seems an odd formulation. It auto-adds them?

`apt show systemtap-sdt-dev` returns this:

This package contains header file and executable (dtrace) that
 can be used to add static probes into userspace applications
 for use with SystemTap.

**Comment by jonatack** [comment]
Thanks @0xB10C, nice work! A few minor comments above while reading the rendered version in localhost.

File: _posts/2021-05-26-#22006.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+- Issue [#20981] discusses potential tracepoints.
+
+[systemtaps]: https://sourceware.org/systemtap/wiki
+[#19866]: https://github.com/bitcoin/bitcoin/pull/19866
+[#20981]: https://github.com/bitcoin/bitcoin/issues/20981
+
+### Tracepoints and tracing scripts
+
+- We can hook into the tracepoints with tracing scripts via the Linux kernel.
+- If we don't hook into the tracepoints, then they are [NOP]s and have little to no performance impact.
```
Comment (jonatack) [inline _posts/2021-05-26-#22006.md:28]:
```suggestion
- If we don't hook into the tracepoints, then they are [NOPs] and have little to no performance impact.
```

File: _posts/2021-05-26-#22006.md (hunk 6 above)
Comment (0xB10C) [reply _posts/2021-05-26-#22006.md:81]: I'm changing this to

```suggestion
- For building Bitcoin Core with USDT support, you need the `sys/std.h` headers (when present, USDT support is [automatically] compiled in). On Debian-like systems you can install the package `systemtap-sdt-dev` (this is not documented yet).
```

**Comment by 0xB10C** [comment]
Appreciate your comments, thanks @jonatack!

**Comment by jonatack** [comment]
Thanks @0xB10C! Do you plan to add the questions here today?

**Comment by 0xB10C** [comment]
No, likely tomorrow in a separate PR.

**Comment by 0xB10C** [comment]
Pushed questions :slightly_smiling_face:

**Comment by jonatack** [comment]
Nice! Back ASAP as I'm in a seminar study group meeting.

**Comment by jonatack** [comment]
Thanks @0xB10C! Made some minor spelling and rendering fixes. Looking forward to learning more on this topic!

File: _posts/2021-05-26-#22006.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+- [collabora.com: An eBPF overview, part 2: Machine & bytecode](https://www.collabora.com/news-and-blog/blog/2019/04/15/an-ebpf-overview-part-2-machine-and-bytecode/)
+- [Brendan D. Gregg's blog posts, and book on on eBPF](http://www.brendangregg.com/)
+- [Brendan D. Gregg: Linux bcc/BPF Node.js USDT Tracing](http://www.brendangregg.com/blog/2016-10-12/linux-bcc-nodejs-usdt.html)
+
+
+
+### Examples
+
+- The PR includes examples and documentation on how to run them.
+- For building Bitcoin Core with USDT support, you need the `sys/std.h` headers (when present, USDT support is [automatically] compiled in). On Debian-like systems you can install the package`systemtap-sdt-dev` (this is not documented yet).
```
Comment (mzumsande) [inline _posts/2021-05-26-#22006.md]: typo: It's sys/s**dt**.h

Comment (jonatack) [reply _posts/2021-05-26-#22006.md]: Thanks @mzumsande, fixed with d3a3b632573811a9e356041c7b7824e39e0318aa.

**Pull request merged**